sdk/azcore/messaging/cloudevent.go

```diff
@@ -0,0 +1,268 @@
+	Extensions map[string]any
+
+	// Subject of the event, in the context of the event producer (identified by Source).
+	Subject *string
+
+	// Time represents the time this event occurred.
+	// Defaults to time.Now().UTC()
+	Time *time.Time
+}
+
+// NewCloudEvent creates a CloudEvent.
+//   - source - Identifies the context in which an event happened. The combination of id and source must be unique
+//     for each distinct event.
+//   - eventType - Type of event related to the originating occurrence.
+//   - data - data to be added to the event. Can be a []byte, or any JSON serializable type, or nil.
+//   - options - additional fields that are not required.
+func NewCloudEvent(source string, eventType string, data any, options *CloudEventOptions) (CloudEvent, error) {
+	id, err := uuid.New()
+
+	if err != nil {
+		return CloudEvent{}, err
+	}
+
+	ce := CloudEvent{
+		ID:          id.String(),
+		Source:      source,
+		SpecVersion: "1.0",
+		Type:        eventType,
+
+		// optional but probably always filled in.
+		Data: data,
+	}
+
+	if options != nil {
+		ce.DataContentType = options.DataContentType
+		ce.DataSchema = options.DataSchema
+		ce.Extensions = options.Extensions
+		ce.Subject = options.Subject
+
+		if options.Time != nil {
+			ce.Time = options.Time
+		}
+	}
+
+	if ce.Time == nil {
+		ce.Time = to.Ptr(time.Now().UTC())
+	}
+
+	return ce, nil
+}
+
+// MarshalJSON implements the json.Marshaler interface for CloudEvent.
+func (ce CloudEvent) MarshalJSON() ([]byte, error) {
+	m := map[string]any{
+		"id":          ce.ID,
+		"source":      ce.Source,
+		"specversion": ce.SpecVersion,
+		"type":        ce.Type,
+	}
+
+	if ce.Data != nil {
+		bytes, isBytes := ce.Data.([]byte)
+
+		if isBytes {
+			m["data_base64"] = base64.StdEncoding.EncodeToString(bytes)
+		} else {
+			m["data"] = ce.Data
+		}
+	}
+
+	if ce.DataContentType != nil {
+		m["datacontenttype"] = ce.DataContentType
+	}
+
+	if ce.DataSchema != nil {
+		m["dataschema"] = ce.DataSchema
+	}
+
+	for k, v := range ce.Extensions {
+		m[k] = v
+	}
+
+	if ce.Subject != nil {
+		m["subject"] = ce.Subject
+	}
+
+	if ce.Time != nil {
+		m["time"] = ce.Time.Format(time.RFC3339Nano)
+	}
+
+	return json.Marshal(m)
+}
+
+func getValue[T any](k string, rawV any, dest *T) error {
+	v, ok := rawV.(T)
+
+	if !ok {
+		var t T
+		return fmt.Errorf("field %q is a %T, but should be %T", k, rawV, t)
+	}
+
+	*dest = v
+	return nil
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface for CloudEvent.
+func (ce *CloudEvent) UnmarshalJSON(data []byte) error {
+	var m map[string]json.RawMessage
+
+	if err := json.Unmarshal(data, &m); err != nil {
+		return err
+	}
+
+	for k, raw := range m {
+		if err := updateFieldFromValue(ce, k, raw); err != nil {
+			return fmt.Errorf("failed to deserialize %q: %w", k, err)
+		}
+	}
+
+	return nil
+}
+
+func updateFieldFromValue(ce *CloudEvent, k string, raw json.RawMessage) error {
+	switch k {
+	//
+	// required attributes
```

Comment:
Should this func return an error if any required attributes are absent?

Reply:
Added (in a few spots)